### DISCUSSION TITLE
First time coding with Python, and I'm trying to make an RC car

### DISCUSSION THREAD

**Discussion by TyMT23** [pybricks/support#1766] [Q&A]
Hi everyone!

Recently I got a technic hub and motors from the RC Audi set and I want to repurpose the electronics to make a custom RC car. I have found a gearbox online on rebrickable, [here,](https://rebrickable.com/mocs/MOC-67989/kueden/4-speed-automatic-gearbox/#details) and want to use this design in my own car. The MOC maker also had instruction on how to repurpose the hub to use pybricks which I had no idea about, until I saw the MOC. I've been browsing and troubleshooting for a few days now and can't seem to get the program to work. Errors that pop up are in lines 21 and 22 saying there's no module named, "math" and "can't import name getchar" and I have no idea how to fix them as I'm still new to pybricks. 

On top of that, I want to use two motors for driving the gearbox instead of one for more torque and speed. This, of course, is mainly a building and space challenge, but I am also struggling to program a second drive motor. It's been a few years since I've coded anything major, and I've never used python, so I'm not quite familiar with everything yet. 

Even more, I want to use an Xbox controller to control my RC car, as shown, [here.](https://pybricks.com/project/technic-42160-xbox/) How should I go about combining both the code for the gearbox, as well as drive it using an Xbox controller. 

My two main issues I need help with, other than needed a crash course in python, is adding a second motor to drive the RC car, and paring that code run with an Xbox controller. Any help is much appreciated, and if anyone has any good advice / videos to give an overview of coding in python that would be amazing as well. <3

Edit: I got my xbox controller to connect! following the instructions in the second link worked perfectly and now all I need is to import the code for the gearbox! My problems from before still persist in the gearbox aspect, "no module named math" and "can't import name getchar" are the bane of my existence at the moment.

Edit2: I asked ChatGPT for help with coding and removing the "getchar" function, and I discovered that the variable isn't called "math" and is instead called "umath" so that problem is solved. For now my problems are, having automatic load detection for automatic gear shifting, and making sure the xbox controller powers the motors with the right buttons. (currently left trigger works for reverse, and right trigger isn't doing anything)

Edit3: There seems to be a "race" mode in the gearbox code, I don't care much for a race mode and would like it removed, I'm a little scared to do so incase it messes with other code. any help on this topic would be much appreciated.

**Comment by BertLindeman** [comment]
The AI assistents like chatGPT are good helpers on python and micropthon. 

Could you share your python program here, so we can have a look?
Please do the markdown around your program 


\```python
enter Your Python code here
\```


A simple hint: if you want to test if a part of the program can be removed, make it a comment:
select that part of the source and press crtl and `/`
To uncomment do the same, so select and press ctrl and `/`

Bert

**Reply by TyMT23** [reply]
Here's the code I have so far. It's currently 3am for me so I'm going to put this to rest for now and look up a basic python coding course tomorrow morning to try and understand the language a bit more. For the time being, here's the code I've got so far. 

```python
from pybricks.hubs import TechnicHub
from pybricks.pupdevices import Motor
from pybricks.parameters import Port, Direction, Color
from pybricks.tools import wait, StopWatch
from pybricks.iodevices import XboxController
from pybricks.robotics import Car

# Initialize the hub, motors, and Xbox controller
hub = TechnicHub()
drive_motor1 = Motor(Port.A)
drive_motor2 = Motor(Port.B)
shift_motor = Motor(Port.C)
steer_motor = Motor(Port.D, Direction.CLOCKWISE)
xbox = XboxController()

# Initialize Car object
car = Car(steer_motor, shift_motor, [drive_motor1, drive_motor2])

# Control settings
stmax = 60
trot = 3000
trace = 5000
stf = 1
stpw = 90

watch = StopWatch()
t0 = watch.time()

# Flags and variables
run = False
rot = 0
gear = 0
gearm = 0
dri = 1

while True:
    # Get Xbox controller inputs
    left_joystick_x, left_joystick_y = xbox.joystick_left()
    right_trigger, left_trigger = xbox.triggers()

    # Control drive and steering based on Xbox controller input
    drive_power = right_trigger - left_trigger
    steer_value = left_joystick_x
    car.steer(steer_value)
    car.drive_power(drive_power)
    
    # Example logic for shifting and race mode (can be customized based on your needs)
    if right_trigger > 0.5:
        # Example logic for shifting or race mode based on controller input
        shift_motor.track_target(gear)  # Set target gear (customize as needed)
        run = True
    else:
        # Stop or adjust gear when no acceleration is applied
        drive_motor1.dc(0)
        drive_motor2.dc(0)
        run = False
    
    # Handle race mode
    if run and (watch.time() - t0 > trace):
        run = False
        gear = 0

    # Adjust steering based on the joystick input
    steer_motor.dc(stf * steer_value)  # Adjust steering power

    wait(50)  # Wait to maintain control loop timing
```

Most of this code was not written by me, and I can't tell what was from the original "gearbox" program and the "xbox controller" program. There is 100% a few bugs in the code and I'm okay with that for now as this isn't a time sensitive project, that being said, I'd love to know if there's any better / smarter way of going about this program.

Thanks so much for your help!

Edit: Forgot to mention, in the original gearbox code there is load detection integrated into the software to help switch gears smoothly. I have no idea how this system functions, and I don't know if ChatGPT included it in the code itself. This is absolutely a necessary feature for this project, and any insight on how this works would be amazing. Thanks again!

**Reply by BertLindeman** [reply]
Have a good night.

Understanding python is a good thing to start with.
Maybe [python course from w3schools](https://www.w3schools.com/python/default.asp) is a start.
It is a `python` course and not all of python is supported in micropython and pybricks. 
But it gives you the base understanding of how python works.
I always forget the differences between the data types and do look that up on w3schhols.

Example | Data Type
-- | -- 
x = "Hello World" | str 
x = 20 | int 
x = 20.5 | float 
x = ["apple", "banana", "cherry"] | list 
x = ("apple", "banana", "cherry") | tuple
x = {"name" : "John", "age" : 36} | dict
x = {"apple", "banana", "cherry"} | set

Bert

[EDIT] The combination Technichub and xbox controller is hard to debug.
An idea might to start using the keyboard to control your car and after that works, replace the keyboard presses by the xbox controller functions?

**Comment by TyMT23** [comment]
Not sure who will see this, but I have some updated code that seems to work somewhat well. Here it is.

```python
from pybricks.pupdevices import Motor
from pybricks.parameters import Port, Stop, Color
from pybricks.tools import wait, StopWatch
from pybricks.hubs import TechnicHub
import umath

# Import required MicroPython libraries.
from usys import stdin
from uselect import poll

# Register the standard input so we can read keyboard presses.
keyboard = poll()
keyboard.register(stdin)

# Initialize motors and hub
drive1 = Motor(Port.C)  # Existing drive motor on Port C
drive2 = Motor(Port.D)  # New drive motor on Port D, works in sync with drive1
shift = Motor(Port.A)
hub = TechnicHub()

# Initialize variables as in your original code
watch = StopWatch()
t = watch.time()
t0 = watch.time()

pst = 0
st0 = 0
stmax = 60
trot = 3000
strot = stmax - 5
trace = 5000
race = False
ready = False
rrl = [1.2, 1.275, 1.35, 1.35]
rrll = [0.6, 0.6, 0.6, 0.75]
gears = [290, 190, 100, 10]
spwr = [-60, -60, -90, -90]
dpwr = 10
g0 = 30
ta = 100
x = 0
rr = 0
rr0 = 0
pwr = 0
y = 0
stf = 1
stpw = 90
new = True
key = None
nn = 100
ty = 0
rot = 0
gear = 0
gearm = 0
run = False
a0 = drive1.angle()  # Initialize with drive1's angle
dri = 1

# Brings the shifting into the first gear
while t - t0 <= 300:
    shift.track_target(g0)
    t = watch.time()

# Function to implement smooth braking
def smooth_brake():
    global pwr, dri, gear, gearm, run
    braking_power = pwr
    while abs(braking_power) > 0:
        # Gradually reduce the power
        braking_power += 5 * dri  # Adjust increment for smoother braking
        if abs(braking_power) < 5:
            braking_power = 0  # Stop when power is very low

        drive1.dc(braking_power)
        drive2.dc(braking_power)  # Sync the second motor

        wait(100)  # Adjust wait time for smoother braking

    # Stop the motors completely
    drive1.dc(0)
    drive2.dc(0)
    pwr = 0
    run = False

# Main loop
while True:
    # Check if a key has been pressed
    if keyboard.poll(0):
        # Read the key and handle it
        key = stdin.read(1)
        print("You pressed:", key)

        # Drive backward
        if key == 'a':
            if new and not run:
                gear = 0
                gearm = 0
                dri = -1
                run = True

            if new and gearm > 0:
                gear = gear - 1
                gearm = gearm - 1
                new = False

            nn = 0
            rot = 0
            stf = 1
            race = False

        # Drive forward
        elif key == 'd':
            if new and gearm < 3:
                if run:
                    gearm = gearm + 1
                    dri = 1
                new = False

            dri = 1
            nn = 0
            rot = 0
            run = True
            stf = 1
            race = False

        # Smooth Brake
        elif key == 's':
            smooth_brake()

        # Kick down acceleration
        elif key == 'w':
            gearm = 3
            dri = 1
            rot = 0
            run = True
            stf = 1
            race = False

        # Start race mode
        elif key == ' ':
            gearm = 3
            dri = 1
            rot = 0
            run = True
            stf = 1
            race = True
            ready = True

        # Steer right
        elif key == 'l':
            if new:
                y = stmax
                new = False

            nn = 0
            rot = 0
            stf = 1

        # Steer left
        elif key == 'j':
            if new:
                y = -stmax
                new = False

            nn = 0
            rot = 0
            stf = 1

        # Center steering
        elif key == 'k':
            y = 0
            rot = 0

        # Tune steering left
        elif key == 'u':
            st0 = st0 - 1
            hub.light.on(Color.GREEN)
            stf = 5

        # Tune steering right
        elif key == 'o':
            st0 = st0 + 1
            hub.light.on(Color.GREEN)
            stf = 5

        # Rotate right
        elif key == 'q':
            rot = 1
            dir = 1
            gearm = 0
            gear = 0
            run = False
            x = 0
            y = 0
            stf = 1

        # Rotate left
        elif key == 'e':
            rot = -1
            gearm = 0
            gear = 0
            run = False
            dir = 1
            x = 0
            y = 0
            stf = 1

    # Rest of your existing code handling motor controls
    # execution of routine for rotating left or right, etc.

    # Update motor positions, power, and other conditions
    # ...

    # Steer angle adjustment and motor control
    pw = (st0 + y - pst) * stf
    if abs(pw) > stpw:
        pw = pw * stpw / abs(pw)
    # steer.dc(pw)

    # Update status light on hub
    if t - ty > 100:
        ty = watch.time()
        hub.light.off()

    # Set the gear shift angle
    an = gears[gear] + g0
    shift.track_target(an)
    t = watch.time()
    dt = t - t0

    if race:
        if ready:
            trace0 = watch.time()
            ready = False

        if t - trace0 >= trace:
            run = False
            race = False
            gear = 0

    # Drive operations (negative power causes forward driving)
    if run:
        if dt >= ta:
            rr = (drive1.angle() - a0) / dt
            t0 = watch.time()
            a0 = drive1.angle()
            if pwr > -100:
                pwr = (pwr - dpwr)
            else:
                pwr = -100

            if abs(rr * 100 / pwr) <= rrll[gear] and rr - rr0 >= 0 and abs(pwr) > 50:
                if gear != 0:
                    gear = gear - 1
                    pwr = spwr[gear]
                    hub.light.on(Color.RED)
                    rr0 = 0
                else:
                    run = False
                    pwr = 0
                    rr0 = 0

            if abs(rr) >= rrl[gear] and gear < gearm:
                gear = gear + 1
                pwr = spwr[gear]
                hub.light.on(Color.BLUE)
                rr0 = 0

        # Synchronize both drive motors
        drive1.dc(dri * pwr)
        drive2.dc(dri * pwr)  # Add this line to sync drive2 with drive1
        rr0 = rr
    else:
        drive1.dc(0)
        drive2.dc(0)  # Stop drive2 when drive1 stops
        rr = 0
        rr0 = 0
```

**Reply by BertLindeman** [reply]
[EDIT] rephrased a bit:
If you edit this post and change the first 3 single quotes \''' into 3 backticks \``` like this: \```python 
and the final 3 single quotes \''' into 3 backticks \``` it will look like python and can simply be copied to test with it.

**Comment by laurensvalk** [comment]
> I have found a gearbox online on rebrickable (...) and want to use this design in my own car. The MOC maker also had instruction on how to repurpose the hub to use pybricks which I had no idea about, until I saw the MOC. I've been browsing and troubleshooting for a few days now and can't seem to get the program to work.

If you're new to programming, adapting someone else's large program can be a bit daunting. You may find it easier to start out with block coding or with a smaller program. You can find example projects with the Technic Hub and the Xbox Controller here:

https://pybricks.com/project/xbox-controller/

This includes both block programs and Python programs. The bulldozer also has a function selector that may act as good inspiration for controlling a gearbox. You could also start without a gearbox and use some of the car examples to get familiar with driving around.
